contrib/babelfishpg_common/src/datetime.c

```diff
@@ -92,20 +434,56 @@ datetime_in_str(char *str, Node *escontext)
 		PG_RETURN_TIMESTAMP(result);
 	}
 
-	dterr = ParseDateTime(str, workbuf, sizeof(workbuf),
+	tm->tm_year = 0;
+	tm->tm_mon = 0;
+	tm->tm_mday = 0;
+
+	strcpy(modified_str, str);
+
+	modified_str = clean_input_str(modified_str, &contains_extra_spaces, DATE_TIME);
+
+	dterr = ParseDateTime(modified_str, workbuf, sizeof(workbuf),
 						  field, ftype, MAXDATEFIELDS, &nf);
+
+	/*
+	 * throw error if the return value is not zero.
+	 */
+	if(tsql_decode_datetime_fields(str, modified_str, field, nf, ftype, 
+					contains_extra_spaces, tm, &is_year_set))
+	{
+		if (modified_str)
+			pfree(modified_str);
+
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_DATETIME_FORMAT),
+				errmsg("invalid input syntax for type datetime: \"%s\"", str)));
+	}
+
+	if (modified_str)
+		pfree(modified_str);
+	
 	if (dterr == 0)
 		dterr = DecodeDateTime(field, ftype, nf, 
-							   &dtype, tm, &fsec, &tz, &extra);
+							&dtype, tm, &fsec, &tz, &extra);
+
 	/* dterr == 1 means that input is TIME format(e.g 12:34:59.123) */
-	/* initialize other necessary date parts and accept input format */
-	if (dterr == 1)
+	if (dterr == 1 || is_year_set)
+	{
+		if (!is_year_set)
+			tm->tm_year = 1900;
+		if (!tm->tm_mon)
+			tm->tm_mon = 1;
+		if (is_year_set || !tm->tm_mday)
+			tm->tm_mday = 1;
+		dterr = 0;
+	}
+	else if(dterr == 1)
 	{
```

Comment:
I think we can keep the initial `if (dterr == 1)` condition and add additional condition for `is_year_set` inside it.